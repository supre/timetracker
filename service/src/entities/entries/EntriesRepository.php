<?php


namespace RoarProj\entities\entries;


use Doctrine\ORM\EntityRepository;

class EntriesRepository extends EntityRepository
{

}