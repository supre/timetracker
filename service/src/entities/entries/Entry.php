<?php
/** @noinspection PhpUnusedAliasInspection */


namespace RoarProj\entities\entries;


use DateTime;
use RoarProj\entities\user\User;
use Doctrine\ORM\Mapping as ORM;

/**
 * Class Emtry
 *
 * @package ts\entities\entries
 *
 * @ORM\Entity(repositoryClass="RoarProj\entities\entries\EntriesRepository")
 * @ORM\Table(
 *     name="entries"
 *     )
 */
class Entry
{
    public function __construct(DateTime $date, ?string $notes, ?float $hoursWorked)
    {
        $this->date = $date;
        $this->notes = $notes;
        $this->hoursWorked = $hoursWorked;
    }

    /**
     * @return int
     */
    public function getId()
    : int
    {
        return $this->id;
    }

    /**
     * @return DateTime
     */
    public function getDate()
    : DateTime
    {
        return $this->date;
    }

    /**
     * @param DateTime $date
     * @return bool
     */
    public function dateBeforeOrEqual(DateTime $date): bool {
        return $this->getDate() <= $date;
    }

    /**
     * @param DateTime $date
     * @return bool
     */
    public function dateAfterOrEqual(DateTime $date): bool {
        return $this->getDate() >= $date;
    }

    /**
     * @return string
     */
    public function getNotes()
    : string
    {
        return $this->notes;
    }

    public function setNotes(string $notes)
    : void {
        $this->notes = $notes;
    }

    /**
     * @return float
     */
    public function getHoursWorked()
    : float
    {
        return $this->hoursWorked;
    }

    public function setHoursWorked(float $hoursWorked)
    : void {
        $this->hoursWorked = $hoursWorked;
    }

    public function setUser(User $owner)
    {
        $this->user = $owner;
    }

    public function unsetUser()
    {
        $this->user = null;
    }

    /**
     * @var integer
     * @ORM\Id()
     * @ORM\Column(type="integer")
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @var DateTime
     *
     * @ORM\Column(
     *     type="datetime",
     * )
     */
    private $date;

    /**
     * @var string
     *
     * @ORM\Column(
     *     type="text",
     * )
     */
    private $notes;

    /**
     * @var float
     *
     * @ORM\Column(
     *     name="hours_worked",
     *     type="float",
     *     length=5)
     */
    private $hoursWorked;

    /**
     * @var User
     *
     * @ORM\ManyToOne(
     *     targetEntity="RoarProj\entities\user\User",
     *     inversedBy="entries"
     * )
     * @ORM\JoinColumn(
     *     name="user_id",
     *     referencedColumnName="id",
     *     nullable=false,
     *     onDelete="CASCADE"
     * )
     */
    private $user;
}