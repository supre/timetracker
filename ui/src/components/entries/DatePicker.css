.clndr-calendars-wrapper,
.clndr-calendars-wrapper > a:not(.ui) {
  color: #000;
}

.entryModal div.field {
  display: inherit !important;
  width: 100%;
}
