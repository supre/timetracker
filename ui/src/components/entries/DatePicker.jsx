import React from "react";
import SemanticDatepicker from "react-semantic-ui-datepickers";
import "react-semantic-ui-datepickers/dist/react-semantic-ui-datepickers.css";
import "./DatePicker.css";

export default ({
  onDateChange,
  placeholder,
  value,
  type = "range",
  pointing = "right",
  iconPosition = "right",
  loading = false,
}) => (
  <SemanticDatepicker
    datePickerOnly
    onChange={(e, d) => onDateChange(d)}
    placeholder={placeholder}
    type={type}
    allowOnlyNumbers
    pointing={pointing}
    iconPosition={iconPosition === "left" ? iconPosition : null}
    maxDate={new Date()}
    value={value}
    loading={loading}
    disabled={loading}
  />
);
