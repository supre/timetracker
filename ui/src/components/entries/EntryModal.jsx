import React, { useState } from "react";
import {
  Modal,
  Icon,
  Form,
  Input,
  Button,
  TextArea,
  Message,
} from "semantic-ui-react";
import DatePicker from "./DatePicker";
import { formatDateToAtom } from "../../modules/Utilities";
import ServiceClient from "./../../modules/ServiceClient";

export default ({
  openModal,
  onCloseModal,
  addEntryToList,
  currentUser,
  userId,
  entries,
  entryId,
  date = "",
  hoursWorked = "",
  notes = "",
}) => {
  const initialFilters = {
    date: date,
    hoursWorked: hoursWorked,
    notes: notes,
  };

  const [filters, setfilters] = useState(initialFilters);
  const [loading, setloading] = useState(false);
  const [id, setId] = useState(entryId);
  const serviceClient = new ServiceClient();
  const [errors, setErrors] = useState([]);

  const handleHoursWorkedValidation = () => {
    const { hoursWorked } = filters;
    if (Number(hoursWorked) > 24) {
      setErrors([
        "You can't work for more than 24 hours. You are not THAT productive. Nice try though.",
      ]);

      return false;
    }

    if (Number(hoursWorked) < 0.1) {
      setErrors([
        "Did you simply forget to fill in worked hours, or were you slacking all day?",
      ]);

      return false;
    }

    return true;
  };

  const emptyFormValidation = () => {
    const { date, hoursWorked, notes } = filters;
    if (date.length === 0 || hoursWorked.length === 0 || notes.length === 0) {
      setErrors(["All fields are mandatory"]);
      return false;
    }

    return true;
  };

  const validateForm = () => {
    return emptyFormValidation() && handleHoursWorkedValidation();
  };

  const onDateChange = (d) => {
    const filteredEntries = entries.filter((e) => e.userId === userId);
    const userEntries =
      filteredEntries.length > 0 ? filteredEntries[0].entries : [];
    const entriesOnDate = userEntries.filter(
      (e) => formatDateToAtom(e.date) === formatDateToAtom(d.value)
    );
    if (entriesOnDate.length > 0) {
      const entry = entriesOnDate[0];
      const { id, date, hoursWorked, notes } = entry;
      setfilters({
        date: date,
        hoursWorked: hoursWorked,
        notes: notes,
      });
      setId(id);
    } else {
      setFilter("date", d.value);
    }
  };

  const onClose = () => {
    setfilters(initialFilters);
    setId(null);
    setErrors([]);
    onCloseModal();
  };

  const handleSubmit = () => {
    if (!validateForm()) return;
    setloading(true);
    setErrors([]);

    const { date, hoursWorked, notes } = filters;
    const entry = {
      date: date,
      hoursWorked: hoursWorked,
      notes: notes,
      id: id,
    };

    const request = id
      ? serviceClient.updateEntry(currentUser.team, userId, entry)
      : serviceClient.addEntry(currentUser.team, userId, entry);

    request
      .then((receivedEntry) => {
        addEntryToList(userId, receivedEntry);
        setloading(false);
        onClose();
      })
      .catch((err) => {
        try {
          setErrors([...err]);
        } catch (error) {
          console.log(error);
          setErrors("Something went wrong. Try again or contact admin");
        }
        setloading(false);
      });
  };

  const setFilter = (key, value) => {
    setfilters({ ...filters, [key]: value });
  };

  const renderErrors = () => {
    return errors.map((err) => (
      <Message error>
        <Message.Header>{err}</Message.Header>
      </Message>
    ));
  };

  return (
    <Modal open={openModal} onClose={onClose} basic closeOnDimmerClick={false}>
      <Modal.Header>
        <Icon name="clock" />
        Track your time
      </Modal.Header>

      <Modal.Content>
        <Form onSubmit={handleSubmit} className="entryModal">
          <DatePicker
            onDateChange={(d) => onDateChange(d)}
            type="basic"
            placeholder="Pick date"
            iconPosition="left"
            value={filters.date}
            loading={loading}
          />
          <Form.Field>
            <Input
              disabled={loading}
              fluid
              label="HoursWorked"
              name="hoursWorked"
              value={filters.hoursWorked}
              type="text"
              onChange={(e) => setFilter("hoursWorked", e.target.value)}
            />
          </Form.Field>
          <Form.Field>
            <TextArea
              disabled={loading}
              placeholder="Write your notes..."
              name="notes"
              value={filters.notes}
              onChange={(e) => setFilter("notes", e.target.value)}
            />
          </Form.Field>
        </Form>
      </Modal.Content>
      <Modal.Content>{renderErrors()}</Modal.Content>

      <Modal.Actions>
        <Button
          color="green"
          onClick={handleSubmit}
          loading={loading}
          disabled={loading}
        >
          <Icon name="checkmark" /> Add
        </Button>
        <Button color="red" onClick={onClose} disabled={loading}>
          <Icon name="remove" /> Cancel
        </Button>
      </Modal.Actions>
    </Modal>
  );
};
