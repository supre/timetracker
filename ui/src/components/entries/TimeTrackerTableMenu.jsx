import React from "react";
import { Icon, Menu, Button } from "semantic-ui-react";
import DatePicker from "./DatePicker";
import { useState } from "react";
import EntryModal from "./EntryModal";

export default ({
  userId,
  onRefresh,
  onDateChange,
  onFilterClick,
  isFiltered,
  filterUrl,
  addEntryToList,
  entries,
  currentUser,
}) => {
  const onFiltersSet = (d) => {
    onDateChange(d);
  };

  const [trackModelOpen, settrackModelOpen] = useState(false);

  return (
    <Menu attached="top" compact secondary inverted>
      <Menu.Menu position="right">
        <Menu.Item>
          <Button
            icon="refresh"
            color="facebook"
            onClick={onRefresh}
            inverted
            circular
          />
        </Menu.Item>
        <Menu.Item>
          <EntryModal
            openModal={trackModelOpen}
            onCloseModal={() => settrackModelOpen(false)}
            userId={userId}
            entries={entries}
            addEntryToList={addEntryToList}
            currentUser={currentUser}
          />
          <Button
            content="Track Time"
            icon="clock outline"
            labelPosition="right"
            color="facebook"
            onClick={() => settrackModelOpen(true)}
            inverted
          />
        </Menu.Item>
        <Menu.Item>
          <DatePicker placeholder="Select Range" onDateChange={onFiltersSet} />
          <Button
            icon
            attached="right"
            color="facebook"
            onClick={onFilterClick}
          >
            <Icon name="filter" />
          </Button>
        </Menu.Item>
        <Menu.Item>
          <Button
            as="a"
            href={filterUrl}
            icon="download"
            content="HTML"
            labelPosition="left"
            color="facebook"
            target="_BLANK"
            disabled={!isFiltered}
          />
        </Menu.Item>
      </Menu.Menu>
    </Menu>
  );
};
