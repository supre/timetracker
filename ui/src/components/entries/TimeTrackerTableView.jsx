import React, { useState } from "react";
import { Table, Button, Modal, Message } from "semantic-ui-react";
import EntryModal from "./EntryModal";
import ServiceClient from "./../../modules/ServiceClient";

export default ({
  userId,
  entries,
  addEntryToList,
  deleteEntryFromList,
  currentUser,
}) => {
  const [errors, setErrors] = useState([]);

  const parseDate = (dateInIsoFormat) => {
    const date = new Date(dateInIsoFormat);
    return date.toDateString();
  };

  const DeleteModal = ({ entry, openModal, onCloseModal }) => {
    const [loading, setLoading] = useState(false);
    const serviceClient = new ServiceClient();

    const deleteEntry = (e) => {
      setLoading(true);
      serviceClient
        .deleteEntry(currentUser.team, userId, entry)
        .then(() => {
          deleteEntryFromList(userId, entry);
          setLoading(false);
          onCloseModal();
        })
        .catch((err) => {
          try {
            setErrors(err);
          } catch (e) {
            console.log(e);
            setErrors(["Something went wrong, try again!!!"]);
          }
          setLoading(false);
        });
    };

    return (
      <Modal open={openModal} onClose={onCloseModal}>
        <Modal.Header>Remove an entry</Modal.Header>
        <Modal.Content>
          <p>Are you sure you want to delete this entry</p>
        </Modal.Content>
        <Modal.Content>
          {errors.map((error) => (
            <Message error>{error}</Message>
          ))}
        </Modal.Content>
        <Modal.Actions>
          <Button
            loading={loading}
            disabled={loading}
            positive
            icon="checkmark"
            labelPosition="right"
            content="Yes"
            onClick={deleteEntry}
          />
        </Modal.Actions>
      </Modal>
    );
  };

  const TableEntry = ({ entry }) => {
    const [isDeleteModalOpen, setDeleteModalState] = useState(false);
    const [trackModelOpen, settrackModelOpen] = useState(false);
    const { id, date, hoursWorked, notes } = entry;
    const didUserWorkMore =
      Number(entry.hoursWorked) >= Number(currentUser.preferredWorkingHours);

    return (
      <Table.Row negative={!didUserWorkMore} positive={didUserWorkMore}>
        <Table.Cell>{parseDate(entry.date)}</Table.Cell>
        <Table.Cell>{entry.hoursWorked}</Table.Cell>
        <Table.Cell>{entry.notes}</Table.Cell>
        <Table.Cell>
          <EntryModal
            openModal={trackModelOpen}
            onCloseModal={() => settrackModelOpen(false)}
            userId={userId}
            entries={entries}
            addEntryToList={addEntryToList}
            date={date}
            notes={notes}
            hoursWorked={hoursWorked}
            currentUser={currentUser}
            entryId={id}
          />
          <Button
            circular
            icon="edit"
            basic
            onClick={() => settrackModelOpen(true)}
          />
          <DeleteModal
            openModal={isDeleteModalOpen}
            onCloseModal={() => setDeleteModalState(false)}
            entry={entry}
          />
          <Button
            circular
            icon="delete"
            basic
            onClick={() => setDeleteModalState(true)}
          />
        </Table.Cell>
      </Table.Row>
    );
  };

  const loadEntries = () => {
    return entries.map((entry, i) => <TableEntry key={i + 1} entry={entry} />);
  };

  return (
    <Table attached="bottom">
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell width={3}>Date</Table.HeaderCell>
          <Table.HeaderCell width={3}>Hours Worked</Table.HeaderCell>
          <Table.HeaderCell width={8}>Notes</Table.HeaderCell>
          <Table.HeaderCell width={2}>Action</Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      {entries.length > 0 ? (
        <Table.Body>{loadEntries()}</Table.Body>
      ) : (
        <Table.Body>
          <Table.Row>
            <Table.Cell colspan={4} style={{ textAlign: "center" }}>
              <Message>
                You don't have any entries. Start tracking your time now.
              </Message>
            </Table.Cell>
          </Table.Row>
        </Table.Body>
      )}
    </Table>
  );
};
