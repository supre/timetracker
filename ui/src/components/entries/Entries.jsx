import React, { useState, useEffect } from "react";
import TimeTrackerTableMenu from "./TimeTrackerTableMenu";
import TimeTrackerTableView from "./TimeTrackerTableView";
import { Segment, Message } from "semantic-ui-react";
import { connect } from "react-redux";
import {
  loadEntries,
  addEntryToList,
  deleteEntryFromList,
} from "../../store/Actions";
import { withRouter } from "react-router-dom";
import ServiceClient from "./../../modules/ServiceClient";

export const Entries = ({
  entries,
  loadEntries,
  addEntryToList,
  currentUser,
  match,
  deleteEntryFromList,
}) => {
  const [userLoading, setUserLoading] = useState(true);
  const [entriesLoading, setEntriesLoading] = useState(true);
  const [error, setError] = useState(null);
  const [userId, setUserId] = useState(null);
  const [filter, setFilter] = useState({ before: null, after: null });
  const [isFiltered, setIsFiltered] = useState(false);
  const [isRefresh, setIsRefresh] = useState(false);
  const filteredEntries = entries.filter((entry) => entry.userId === userId)[0];
  const userEntries = filteredEntries && filteredEntries.entries;

  useEffect(() => {
    setUserId((match && match.params && match.params.userId) || currentUser.id);
    setUserLoading(false);
  }, [match, currentUser]);

  useEffect(() => {
    if (!userId) return;
    setError(null);
    const serviceClient = new ServiceClient();
    if (!userEntries || isRefresh) {
      const { before, after } = filter;
      serviceClient
        .getEntries(currentUser.team, userId, before, after)
        .then((entries) => {
          loadEntries(userId, entries);
          setEntriesLoading(false);
          setIsRefresh(false);
          before && after ? setIsFiltered(true) : setIsFiltered(false);
        })
        .catch((err) => {
          setError(err);
          setEntriesLoading(false);
          setIsRefresh(false);
        });
    } else {
      setEntriesLoading(false);
    }
  }, [userId, currentUser, isRefresh, filter]);

  const refreshEntries = () => {
    setEntriesLoading(true);
    setIsRefresh(true);
  };

  const onDateChange = (d) => {
    if (!d.value || d.value.length < 2) {
      setFilter({ before: null, after: null });
    } else {
      setFilter({ after: d.value[0], before: d.value[1] });
    }
  };

  const loadFilteredEntries = () => {
    refreshEntries();
  };

  const downloadUrl = () => {
    if (!isFiltered) return "";

    const { before, after } = filter;
    const url = new ServiceClient().getDownloadUrl(
      currentUser.team,
      userId,
      before,
      after
    );

    return url;
  };

  return (
    <Segment basic loading={userLoading || entriesLoading}>
      <TimeTrackerTableMenu
        onRefresh={refreshEntries}
        onDateChange={onDateChange}
        isFiltered={isFiltered}
        onFilterClick={loadFilteredEntries}
        filterUrl={downloadUrl()}
        currentUser={currentUser}
        entries={entries}
        addEntryToList={addEntryToList}
        userId={userId}
      />
      {error && <Message error>{error}</Message>}
      {!error && !userLoading && !entriesLoading && (
        <TimeTrackerTableView
          userId={userId}
          entries={userEntries ? userEntries : []}
          addEntryToList={addEntryToList}
          deleteEntryFromList={deleteEntryFromList}
          currentUser={currentUser}
        />
      )}
    </Segment>
  );
};

export default withRouter(
  connect(
    (state) => ({
      entries: state.entries,
      currentUser: state.user.currentUser,
    }),
    { loadEntries, addEntryToList, deleteEntryFromList }
  )(Entries)
);
