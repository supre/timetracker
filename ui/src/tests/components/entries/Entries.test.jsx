import React from "react";
import renderer from "react-test-renderer";
import { Entries } from "./../../../components/entries/Entries";

test("render entries", () => {
  let entries = renderer.create(<Entries entries={[]} />);
  let tree = entries.toJSON();
  expect(tree).toMatchSnapshot();
});
