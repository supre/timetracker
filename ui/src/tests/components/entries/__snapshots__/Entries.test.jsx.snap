// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render entries 1`] = `
<div
  className="ui basic loading segment"
>
  <div
    className="ui compact inverted secondary top attached menu"
  >
    <div
      className="right menu"
    >
      <div
        className="item"
        onClick={[Function]}
      >
        <button
          className="ui facebook circular icon inverted button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="refresh icon"
            onClick={[Function]}
          />
        </button>
      </div>
      <div
        className="item"
        onClick={[Function]}
      >
        <button
          className="ui facebook icon inverted right labeled button"
          onClick={[Function]}
        >
          <i
            aria-hidden="true"
            className="clock outline icon"
            onClick={[Function]}
          />
          Track Time
        </button>
      </div>
      <div
        className="item"
        onClick={[Function]}
      >
        <div
          className="field"
          style={
            Object {
              "display": "inline-block",
              "position": "relative",
            }
          }
        >
          <div
            className="field"
          >
            <div
              className="ui icon input"
              data-testid="datepicker-input"
            >
              <input
                disabled={false}
                onBlur={[Function]}
                onChange={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                placeholder="Select Range"
                readOnly={true}
                required={false}
                type="text"
                value=""
              />
              <i
                aria-hidden="true"
                className="calendar link icon"
                data-testid="datepicker-icon"
                onClick={[Function]}
              />
            </div>
          </div>
        </div>
        <div
          className="ui facebook icon right attached button"
          onClick={[Function]}
          role="button"
          tabIndex={0}
        >
          <i
            aria-hidden="true"
            className="filter icon"
            onClick={[Function]}
          />
        </div>
      </div>
      <div
        className="item"
        onClick={[Function]}
      >
        <a
          className="ui facebook icon disabled left labeled button"
          href=""
          onClick={[Function]}
          role="button"
          tabIndex={-1}
          target="_BLANK"
        >
          <i
            aria-hidden="true"
            className="download icon"
            onClick={[Function]}
          />
          HTML
        </a>
      </div>
    </div>
  </div>
</div>
`;
